### PR TITLE
System Job for cleaning up ObjectChange records

### LINKED ISSUE #2352: Deadlocks in long standing Jobs
### Environment
* Nautobot version (Docker tag too if applicable): 1.4.1
* Python version:
* Database platform, version: Postgres
* Middleware(s):

### Steps to Reproduce
1. Having a big Changelog table (>11GB), and having activated the retention policy clean-up (90 days by default), in long standing Jobs, we get randomly generated deadlocks for `extras_objectchange`
2.
3.

### Expected Behavior
No issues with database deadlocks

### Observed Behavior
Deadlocks related to `extras_objectchange` are raised.

### More info
A way to mitigate this has been to disable the retention policy clean-up (setting `CHANGLOG_RETENTION` to 0), and create a specific Job to do the clean-up in a controlled way.
Maybe this [code](https://github.com/nautobot/nautobot/blob/9b2e2dbfe94dd4ef9bec17a414e070b38e4db7f7/nautobot/extras/signals.py#L98-L102) is the responsible of this random behaviour, and it would be the root cause to fix it.

### LINKED ISSUE #3749: Retention Cleanup as Jobs
### As ...

Patti - Platform Admin

### I want ...

Have a built-in job to clean up old `JobLogEntry` and `ObjectChange` records instead of settings.

### So that ...

I...
- Can schedule the cleanups for more opportune times (avoiding things #2352)
- Don't have to re-deploy my config to change my retention settings
- Don't have to drop into the ORM for one-off cleanups
- Have a pattern I can mimic if I want to refine my cleanup further (by types, etc)

### I know this is done when...

Job(s) exist in the Jobs list view I can run and/or schedule to clean up my logs.

Could eliminate #2865, #2501, #2352, #2006, #3678.

### Optional - Feature groups this request pertains to.

- Automation
- Other (not directly a platform feature)

### Database Changes

_No response_

### External Dependencies

_No response_

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes #2352, closes #3749
# What's Changed

- Add ~`ObjectChangeCleanup`~ `LogsCleanup` system Job. Fixes #3749.
  - [x] Need to add this to the Release Overview documentation
  - Takes one argument for now, which is the age cutoff in days to apply. If unspecified, defaults to the value of `CHANGELOG_RETENTION`.
  - To improve performance of the bulk delete, it disables the general-purpose deletion signal `_handle_deleted_objects` as it isn't relevant to the deletion of ObjectChange or JobResult records.
  - Additionally, I reworked the `invalidate_models_cache` signal handler so that it no longer triggers on *all* model saves and deletes, but now only on CustomField/ComputedField/Relationship changes as it was intended to.
  - In some quick profiling, I found this set of changes improved the job runtime by roughly an order of magnitude (e.g. deleting 6000 records improved from 13 seconds to roughly 1.3 seconds).
- Added `JobResultFactory`, `JobLogEntryFactory`, and `ObjectChangeFactory` factory classes and added them to `generate_test_data`.
  - [x] Need to verify whether this impacts any unit test behavior.
  - [x] Need to document this in a change fragment
- Removed automatic random inline cleanup of ObjectChanges from the `_handle_changed_object` signal. Fixes #2352.
  - Removed references to `CHANGELOG_RETENTION` in UI templates since it is no longer an automatic given behavior

# Screenshots

Job output:

image

Factory-generated ersatz ObjectChange records:

image

image

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- n/a Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by glennmatthews** [comment]
#3749 also wants a similar job for cleanup of `JobLogEntry` records. Not sure whether it makes more sense to do a separate job for that purpose or try to combine the two into a single job that can do either or both.

**Review by timizuoebideri1 — APPROVED** [review]

File: nautobot/core/tests/test_filters.py [hunk 1]
```diff
@@ -1304,14 +1304,14 @@ def setUp(self):
             parent=self.parent_location_1,
             name="Test Child Location 1",
             location_type=self.lt,
-            asn=1234,
+            asn=12345,
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/tests/test_filters.py:1307]: With the factory changes, I had a record whose address coincidentally included `1234` and it was making the test fail. Making these more specific was a quick and dirty fix.

**Comment by glennmatthews** [comment]
Moving this back to draft while I work on adding JobResult/JobLogEntry cleanup as a supported option.

**Comment by glennmatthews** [comment]
image

File: nautobot/core/jobs/cleanup.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+from nautobot.core.utils.config import get_settings_or_config
+from nautobot.extras.jobs import IntegerVar, Job, MultiChoiceVar
+from nautobot.extras.models import JobResult, ObjectChange
+from nautobot.extras.signals import _handle_deleted_object
+
+name = "System Jobs"
+
+
+CLEANUP_TYPES = (
```
Comment (timizuoebideri1) [inline nautobot/core/jobs/cleanup.py:15]: What about using choices here:

```python
class CleanUpTypes(...):
      ObjectChange = "Change logs"
      ...
```

File: nautobot/core/jobs/cleanup.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if max_age in (None, ""):
+            max_age = get_settings_or_config("CHANGELOG_RETENTION")
+            if max_age == 0:
+                self.logger.warning(
+                    "CHANGELOG_RETENTION setting is set to zero, disabling this Job. "
+                    "If you wish to use this Job to delete records, you must specify a `max_age` value."
+                )
+                return 0
+
+        if "JobResult" in cleanup_types and not self.user.has_perm("extras.delete_jobresult"):
```
Comment (timizuoebideri1) [inline nautobot/core/jobs/cleanup.py:55]: use const choices e.g `CleanUpTypes.JobResult` here and other places?

File: nautobot/core/tests/test_jobs.py [hunk 4]
```diff
@@ -256,3 +260,118 @@ def test_csv_import_bad_row(self):
...
+        """Job should enforce user permissions on the content-types being deleted."""
+        job_result_count = JobResult.objects.count()
+        job_log_entry_count = JobLogEntry.objects.count()
+        object_change_count = ObjectChange.objects.count()
+
+        job_result = create_job_result_and_run_job(
+            "nautobot.core.jobs.cleanup",
+            "LogsCleanup",
+            username=self.user.username,
+            cleanup_types=["JobResult"],
```
Comment (timizuoebideri1) [inline nautobot/core/tests/test_jobs.py:291]: Use `CleanUpType` choices here and below?

**Pull request merged**